### PR TITLE
Implementation of new workload name label

### PR DESCRIPTION
**Pull request by howardjohn**
Implementation for https://github.com/istio/api/pull/3374/

### CODE REVIEW

File: pkg/kube/labels/labels.go [hunk 1]
```diff
@@ -33,6 +36,17 @@ var (
 	}
 )
 
+// WorkloadNameFromWorkloadEntry derives the workload name from a WorkloadEntry
+func WorkloadNameFromWorkloadEntry(name string, annos map[string]string, labels map[string]string) string {
```
Comment (costinm) [inline pkg/kube/labels/labels.go:40]: So is it annotation or label or both ?

Comment (howardjohn) [reply pkg/kube/labels/labels.go:40]: the name thing is a label, the existing 'workload group' annotation is an annotation though. In this PR I mistakenly made the injection template read it as an annotation which is wrong; should be a label. 

Label is selected to align with canonical-service and canonical-revision

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go [hunk 2]
```diff
@@ -179,7 +179,8 @@ func (a *index) workloadEntryWorkloadBuilder(
 			log.Warnf("skipping workload entry %s/%s; DNS Address resolution is not yet implemented", wle.Namespace, wle.Name)
 		} // Else it is an empty address with network set, this is ok
 
-		w.WorkloadName, w.WorkloadType = wle.Name, workloadapi.WorkloadType_POD // XXX(shashankram): HACK to impersonate pod
+		w.WorkloadName = kubelabels.WorkloadNameFromWorkloadEntry(wle.Name, wle.Annotations, wle.Labels)
```
Comment (costinm) [inline pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:182]: Handling WorklogGroup labels or annotations (or both) - no problem.

My concern remains that this creates an annotation-based API for ambient - where injection is not required. I would rather have the pods that are not part of deployment/RS/etc use their name - and not require annotations.

An option is for the WorkloadGroup selector matching such lonely Pod.

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:182]: >  I would rather have the pods that are not part of deployment/RS/etc use their name

Then we will put in our prometheus metrics things like `application-756cd7dff4-ts8mk` and DOS prometheus.

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:182]: There is no current correlation between labels and annotation usage and ambient. There are plenty of existing usages of them, and no reason I see to not use them.

**Review by bleggett — APPROVED** [review]
API and impl makes sense to me.

**Comment by costinm** [comment]
On Mon, Dec 16, 2024 at 9:59 AM John Howard ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go
> <https://github.com/istio/istio/pull/54340#discussion_r1887270987>:
>
> > @@ -179,7 +179,8 @@ func (a *index) workloadEntryWorkloadBuilder(
>  			log.Warnf("skipping workload entry %s/%s; DNS Address resolution is not yet implemented", wle.Namespace, wle.Name)
>  		} // Else it is an empty address with network set, this is ok
>
> -		w.WorkloadName, w.WorkloadType = wle.Name, workloadapi.WorkloadType_POD // XXX(shashankram): HACK to impersonate pod
> +		w.WorkloadName = kubelabels.WorkloadNameFromWorkloadEntry(wle.Name, wle.Annotations, wle.Labels)
>
> I would rather have the pods that are not part of deployment/RS/etc use
> their name
>
> Then we will put in our prometheus metrics things like
> application-756cd7dff4-ts8mk and DOS prometheus.
>
The common case for pods without replica sets is an application creating
them on the fly (like cert-manager).
I don't think we can require all apps that create pods programmatically to
have Istio-specific labels.

The other approach - using a WorkloadGroup selector that may select Pods
created programmatically - doesn't require changes
to each app, and it works for both WorkloadEntry and (standalone) Pods. It
also works for Pods that are part of a Deployment
actually, also without requiring users to change their helm charts. And we
already have the relevant code.

**Review by stevenctl — APPROVED** [review]

**Pull request merged**